Append parameters to target URL

jarrettkong opened the pull request:
Implement and test logic to append any parameters to the target URL.

jarrettkong committed:
Add logic to join url parameters from yaml file to target url arg

Signed-off-by: Jarrett Kong <jarrettkong@me.com>

jarrettkong committed:
Remove extra & from url param interpolation

jarrettkong committed:
Fix error where the url params were replacing the target url rather than appending to it

Signed-off-by: Jarrett Kong <jarrettkong@me.com>

jarrettkong committed:
Unit test joinURLParams function

Signed-off-by: Jarrett Kong <jarrettkong@me.com>

wlynch commented on pkg/hook/hook.go:
```diff
@@ -10,9 +11,10 @@ import (
+	Method  string            `yaml:"method"`
+	Headers http.Header       `yaml:"headers"`
+	Body    string            `yaml:"body"`
+	Params  map[string]string `yaml:"params"`
```
We should use [url.Values](https://godoc.org/net/url#Values) instead. URL params can actually be specified multiple times.

wlynch commented on pkg/hook/hook.go:
```diff
@@ -55,8 +57,19 @@ func (h *Hook) Dump() ([]byte, error) {
 	return yaml.Marshal(h)
 }
 
+func (h *Hook) joinURLParams() string {
```
Once you switch to using [url.Values](https://godoc.org/net/url#Values), you can replace this method with `h.Params.Encode()`

wlynch commented on pkg/hook/hook.go:
```diff
@@ -55,8 +57,19 @@ func (h *Hook) Dump() ([]byte, error) {
+
 // Fire sends an HTTP request to the given target.
 func (h *Hook) Fire(target string) (*http.Response, error) {
+	if len(h.Params) >= 1 {
```
Once you switch to using url.Values, you can remove this.

wlynch commented on pkg/hook/hook.go:
```diff
@@ -55,8 +57,19 @@ func (h *Hook) Dump() ([]byte, error) {
 // Fire sends an HTTP request to the given target.
 func (h *Hook) Fire(target string) (*http.Response, error) {
+	if len(h.Params) >= 1 {
+		target += h.joinURLParams()
```
Set this in `h.toRequest` instead. e.g. something along the lines of https://play.golang.org/p/YgRtsfSVtE5

eddiezane commented:
@jarrettkong Nice job! Would like to see the record logic implemented along with this to close the issue.

jarrettkong committed:
Refactor Hook params to use url.Values instead of map[string]string.

- Removed joinURLParams receiver function and associated test
- Replaced above func w/ url.Params.Encode

Signed-off-by: Jarrett Kong <jarrettkong@me.com>

wlynch reviewed (changes requested):
Looks good! Let's add tests that include parsing both with and without params. In particular, I want to make sure we don't nil panic on the `h.Params` if there are no explicit params defined.

jarrettkong committed:
WIP: Unit test for parsing URL parameters with and without empty params

Signed-off-by: Jarrett Kong <jarrettkong@me.com>

jarrettkong committed:
WIP: Remove extra code from empty param unit test

Signed-off-by: Jarrett Kong <jarrettkong@me.com>

jarrettkong commented:
@wlynch I was able to test and record the parameters. When they are explicitly defined, the test passes, but unspecified params are failing, even with `reflect.DeepEqual()`. The error is `expected params to be map[] got map[]`, which should pass with `DeepEqual` from my understanding. I also broke the empty params test out into it's own function, as I wasn't sure how to change the params without a new request yet.

jarrettkong commented:
@eddiezane if I run `hook fire` with the correct args, I get the correct output from record. However, it's automatically changing the `params` in the `test.yml` file from what is defined there to just `{}` and I can't figure out why. After each run I have to go in and undo it manually - very odd.

eddiezane commented:
@jarrettkong let's rebase with master and take another look

jarrettkong committed:
Merge branch 'master' into jarrettkong/url-parameters

eddiezane committed:
Fix tests

Signed-off-by: Eddie Zaneski <eddiezane@gmail.com>

eddiezane committed:
Add tests for record

Signed-off-by: Eddie Zaneski <eddiezane@gmail.com>

wlynch commented on cmd/record_test.go:
```diff
@@ -55,48 +55,99 @@ func reqString(t *testing.T, req *http.Request) string {
+			t.Fatal(err)
+		}
+
+		for k, v := range tc.headers {
```
I think you can replace with with `req.Header = tc.headers`

eddiezane replied:
Good catch. I was worried that would blow away the default headers but I it makes sense that they are populated last.

eddiezane committed:
Cleanup

Signed-off-by: Eddie Zaneski <eddiezane@gmail.com>
